### PR TITLE
Fix and enable checkpoints for mainnet (WIP)

### PR DESCRIPTION
**Pull request by chjj**
Long overdue PR to enable checkpoints on mainnet. Currently untested.

### CODE REVIEW

File: lib/protocol/networks.js [hunk 1]
```diff
@@ -84,7 +84,23 @@ main.brontidePort = 44806;
...
+  30000: Buffer.from(
+    '0000000000000004f790862846b23c3a81585aea0fa79a7d851b409e027bcaa7', 'hex'),
+  40000: Buffer.from(
+    '0000000000000002966206a40b10a575cb46531253b08dae8e1b356cfa277248', 'hex'),
+  50000: Buffer.from(
+    '00000000000000020c7447e7139feeb90549bfc77a7f18d4ff28f327c04f8d6e', 'hex'),
+  56880: Buffer.from(
+    '0000000000000001d4ef9ea6908bb4eb970d556bd07cbd7d06a634e1cd5bbf4e', 'hex'),
+  61043: Buffer.from(
+    '00000000000000015b84385e0307370f8323420eaa27ef6e407f2d3162f1fd05', 'hex')
```
Comment (Falci) [inline lib/protocol/networks.js:103]: Why these specific blocks?

File: lib/blockchain/chain.js [hunk 2]
```diff
@@ -659,17 +659,18 @@ class Chain extends AsyncEmitter {
...
       const tx = block.txs[i];
 
-      assert(await view.spendInputs(this.db, tx),
-        'BUG: Spent inputs in historical data!');
+      if (i === 0) {
+        assert(view.bits.spend(this.db.field, tx));
+      } else {
+        assert(await view.spendInputs(this.db, tx),
+          'BUG: Spent inputs in historical data!');
+      }
```
Comment (pinheadmz) [inline lib/blockchain/chain.js:673]: Besides the tree root assertion, this change looks like a noop to me -- what was fixed?

File: lib/protocol/networks.js (hunk 1 above)
Comment (pinheadmz) [reply lib/protocol/networks.js:103]: ✅  Confirmed these hashes match these heights in main chain on node synced without checkpoints:

```
$ for h in 1008 2016 10000 20000 30000 40000 50000 56880 61043 ; do hsd-rpc getblockhash $h; done;
0000000000001013c28fa079b545fb805f04c496687799b98e35e83cbbb8953e
0000000000000424ee6c2a5d6e0da5edfc47a4a10328c1792056ee48303c3e40
00000000000001a86811a6f520bf67cefa03207dc84fd315f58153b28694ec51
0000000000000162c7ac70a582256f59c189b5c90d8e9861b3f374ed714c58de
0000000000000004f790862846b23c3a81585aea0fa79a7d851b409e027bcaa7
0000000000000002966206a40b10a575cb46531253b08dae8e1b356cfa277248
00000000000000020c7447e7139feeb90549bfc77a7f18d4ff28f327c04f8d6e
0000000000000001d4ef9ea6908bb4eb970d556bd07cbd7d06a634e1cd5bbf4e
00000000000000015b84385e0307370f8323420eaa27ef6e407f2d3162f1fd05
```

**Comment by pinheadmz** [comment]
Seems to be a good improvement although at this chain height on my 12-core machine the savings was only about 20 minutes:

`hsd --no-wallet --prefix=...`

master branch:
```
Height: 62000 Total block processing time: 1718497.5929899993 (0.47736044249722204 hours)
Total time elapsed: 1742000ms (0.4838888888888889 hours)
```

This PR branch:
```
Height: 62000 Total block processing time: 1577276.6586819978 (0.43813240518944385 hours)
Total time elapsed: 1608000ms (0.44666666666666666 hours)
```

**Comment by pinheadmz** [comment]
Going to merge this now and add tests in https://github.com/handshake-org/hsd/pull/629

**Comment by handshake-enthusiast** [comment]
>Would also appreciate some clear policy doc on how new checkpoints will be added going forward.

It was clarified later in a0aa531df6b7737e5a8910d12a0af91eaa1f8ce7. From [docs/release-process.md#release-candidate-and-major-release](https://github.com/handshake-org/hsd/blob/a0aa531df6b7737e5a8910d12a0af91eaa1f8ce7/docs/release-process.md#release-candidate-and-major-release):  
>Add new checkpoint. Last checkpoint + 6 months.

**Comment by nodech** [comment]
I believe it should be fixed (as the release process is currently not following 100% schedule) to:
  Last checkpoint should be -4~6 months before release. (v6 is taking a while because of soft-fork)

**Comment by handshake-enthusiast** [comment]
I created https://github.com/handshake-org/hsd/pull/827, so you don't forget to change it.

BTW checkpoints are going to be removed from Bitcoin – https://github.com/bitcoin/bitcoin/pull/25725.

**Pull request merged**